dps-common/caasauth/caasauth.go

```diff
@@ -0,0 +1,254 @@
+package caasauth
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/ernesto-jimenez/httplogger"
+	"github.com/go-openapi/runtime"
+	"github.com/go-openapi/runtime/client"
+	"github.com/hortonworks/cb-cli/dps-common/utils"
+)
+
+var PREFIX_TRIM = []string{"http://", "https://"}
+
+// This is nearly identical with http.DefaultTransport
+var TransportConfig = &http.Transport{
+	Proxy:           http.ProxyFromEnvironment,
+	TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	Dial: (&net.Dialer{
+		Timeout:   30 * time.Second,
+		KeepAlive: 30 * time.Second,
+	}).Dial,
+	TLSHandshakeTimeout:   10 * time.Second,
+	ExpectContinueTimeout: 1 * time.Second,
+}
+
+var LoggedTransportConfig = httplogger.NewLoggedTransport(TransportConfig, newLogger())
+
+type tokenRequest struct {
+	GrantType string `json:"grant_type"`
+	Code      string `json:"code"`
+	ClientId  string `json:"client_id"`
+}
+
+type refreshRequest struct {
+	GrantType    string `json:"grant_type"`
+	RefreshToken string `json:"refresh_token"`
+	ClientId     string `json:"client_id"`
+}
+
+func newTokenRequest(code string) *tokenRequest {
+	return &tokenRequest{
+		GrantType: "authorization_code",
+		Code:      code,
+		ClientId:  "6eda2bf3-95ce-499b-8e27-1c19c93bae12",
+	}
+}
+
+func newRefreshTokenRequest(refresh string) *refreshRequest {
+	return &refreshRequest{
+		GrantType:    "refresh_token",
+		RefreshToken: refresh,
+		ClientId:     "6eda2bf3-95ce-499b-8e27-1c19c93bae12",
+	}
+}
+
+type TokenResponse struct {
+	AccessToken  string `json:"access_token"`
+	TokenType    string `json:"token_type"`
+	ExpiresIn    string `json:"expires_in"`
+	RefreshToken string `json:"refresh_token"`
+}
+
+func RefreshAccessToken(address, baseApiPath, refreshToken string) (*Transport, *TokenResponse) {
+	address, basePath := cutAndTrimAddress(address)
+	tokenReq := newRefreshTokenRequest(refreshToken)
+
+	tokens, err := getCaasToken("https://"+address+basePath+"/oidc/token", tokenReq)
+	if err != nil {
+		utils.LogErrorAndExit(err)
+	}
+	cbTransport := &Transport{client.New(address, basePath+baseApiPath, []string{"https"})}
+	cbTransport.Runtime.DefaultAuthentication = client.BearerToken(tokens.AccessToken)
+	cbTransport.Runtime.Transport = LoggedTransportConfig
+	return cbTransport, tokens
+}
+
+func NewRefreshToken(address string) string {
+	_, tokens := NewCaasTransport(address, "")
+	return tokens.RefreshToken
+}
+
+func NewCaasTransport(address, baseApiPath string) (*Transport, *TokenResponse) {
+	address, basePath := cutAndTrimAddress(address)
+
+	caasPath := fmt.Sprintf("https://%[1]s/oidc/authorize?scope=openid dps offline_access&response_type=code&client_id=6eda2bf3-95ce-499b-8e27-1c19c93bae12&redirect_uri=https://%[1]s/caas/cli&state=random-state&nonce=random-nonce", address+basePath)
```

Comment:
Maybe instead of `address+basePath` we can use two parameters for `fmt.Sprintf()`

Reply:
the Sprintf call would be more complicated and less readable and the result would be the same